SARAALERT 904: Case Status Import Validation

hackrm opened the pull request:
# Description
Jira Ticket: SARAALERT-904

Workflow specific import validation for monitorees imported with 'Case Status' value.

# Demo/Screenshots

**Situation 1**: Monitorees are imported in the exposure dashboard with a value other than blank, “Suspect”, “Unknown” or “Not a Case” --> upon import, the error message in the screenshot below appears & import doesn’t proceed
![image](https://user-images.githubusercontent.com/42656458/97753706-94fa1500-1acc-11eb-8aff-ad3c7c6c9b60.png)

**Situation 2**: Cases are imported in the isolation dashboard with a value other than blank, “Probable”, or “Confirmed” --> upon import, the error message in the screenshot below appears & import doesn’t proceed
![image](https://user-images.githubusercontent.com/42656458/97753734-a2170400-1acc-11eb-93c7-94ee2e74e23c.png)

# Important Changes

`app/controllers/import_controller.rb`
- New method that validates enums for workflow specific values
- More robust validation logic for 'Case Status'

`app/helpers/validation_helper.rb`
- Environmental variable for workflow specific enum values (Case Status is currently the only data element in this dict)

`test/system/roles/public_health/dashboard/import_verifier.rb`
- Updated to handle workflow specific fields

`test/system/roles/public_health/public_health_import_export_test.rb`
- Existing tests that were failing now use updated fixtures taking workflow specific fields into account
- New tests for workflow specific field validation

Excel Fixtures
- Updated and created new fixtures to work with existing import tests

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [x] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] IE11

* [ ] While in the Exposure workflow, import monitorees with case statuses of blank, "Suspect", "Unknown", and "Not a Case". Observe that the import succeeds and displays the import snapshots.
* [ ] While in the Exposure workflow, import monitorees with case statuses of "Probable", "Confirmed" and some other erroneous input. Observe that the import fails and displays error messages for each of the improper Case Status value while mentioning the valid inputs.
* [ ] While in the Isolation workflow, import monitorees with case statuses of blank, "Probable", and "Confirmed". Observe that the import succeeds and displays the import snapshots.
* [ ] While in the Isolation workflow, import monitorees with case statuses of "Suspect", "Unknown", "Not a Case", and some other erroneous input. Observe that the import fails and displays error messages for each of the improper Case Status value while mentioning the valid inputs.

hackrm commented on app/controllers/import_controller.rb:
```diff
@@ -274,6 +276,26 @@ def validate_assigned_user(value, row_ind)
+    if workflow == :exposure
+      return NORMALIZED_EXPOSURE_ENUMS[field][normalized_value] if NORMALIZED_EXPOSURE_ENUMS[field].keys.include?(normalized_value)
+
+      # rubocop:disable Layout/LineLength
```
Question for reviewers - The line length for line 287 was triggering Rubocop. Is it temporary disable appropriate to appease rubocop's line length check in this instance?

kylietmo replied:
We typically avoid doing that in the application code (we do it in some places in the tests), so I'd ask you to try and avoid doing so here too if possible (which I think should be do-able).

hackrm replied:
Just pushed an update splitting the line up.

kylietmo commented on app/helpers/validation_helper.rb:
```diff
@@ -92,10 +92,26 @@ module ValidationHelper # rubocop:todo Metrics/ModuleLength
   end
 
+  NORMALIZED_EXPOSURE_ENUMS = VALID_EXPOSURE_ENUMS.transform_values do |values|
+    Hash[values.collect { |value| [value.to_s.downcase.gsub(/[ -.]/, ''), value] }]
```
Given we have this same line in three places now, it probably makes more sense to just break it out into a helper function.

hackrm replied:
Just pushed an update addressing your suggestion.

kylietmo reviewed (commented):
I went down a rabbit-hole thinking about how we can generalize this to work for any workflow specific fields (like symptom onset date), but that starts to get into the idea of requiring fields (which we don't yet do on import), so I think it's probably out of scope of this ticket. 

Tested and other than my one comment below, it looks good.

timwongj reviewed (dismissed):
similar to @kylietmo , I also tried to think of other ways we could maybe generalize some of this logic, but I think that can come when we have to do other workflow specific validations. this looks good to me and great job with updating the system tests!

kylietmo commented:
@hackrm Would you mind rebasing off of `master` before I merge?

hackrm commented:
> 
> 
> @hackrm Would you mind rebasing off of `master` before I merge?

@kylietmo I'll work on that right now.

kylietmo commented:
After discussion, this should now be targeting 1.17.0. So @hackrm can you change the target branch and rebase off of `1.17.0` instead?

hackrm commented:
> After discussion, this should now be targeting 1.17.0. So @hackrm can you change the target branch and rebase off of `1.17.0` instead?

@kylietmo I just changed the target branch and rebased off of it. So it should be all set!
